pkg/scheduler/plugins/coescheduling/core/gang.go

```diff
@@ -0,0 +1,436 @@
+
+	podId := GetNamespaceSplicingName(pod.Namespace, pod.Name)
+	return gang.ChildrenScheduleRoundMap[podId]
+}
+
+func (gang *Gang) getCreateTime() time.Time {
+	gang.lock.Lock()
+	defer gang.lock.Unlock()
+
+	return gang.CreateTime
+}
+
+func (gang *Gang) getGangGroup() []string {
+	gang.lock.Lock()
+	defer gang.lock.Unlock()
+
+	return gang.GangGroup
+}
+
+func (gang *Gang) isGangResourceSatisfied() bool {
+	gang.lock.Lock()
+	defer gang.lock.Unlock()
+
+	return gang.ResourceSatisfied
+}
+
+func (gang *Gang) isGangScheduleCycleValid() bool {
+	gang.lock.Lock()
+	defer gang.lock.Unlock()
+
+	return gang.ScheduleCycleValid
+}
+
+func (gang *Gang) setChild(pod *v1.Pod) {
+	gang.lock.Lock()
+	defer gang.lock.Unlock()
+
+	podId := GetNamespaceSplicingName(pod.Namespace, pod.Name)
+	gang.Children[podId] = pod
+	klog.Infof("SetChild, gangName: %v, childName: %v", gang.Name, podId)
+}
+
+func (gang *Gang) setScheduleCycleValid(valid bool) {
+	gang.lock.Lock()
+	defer gang.lock.Unlock()
+
+	gang.ScheduleCycleValid = valid
+	klog.Infof("SetScheduleCycleValid, gangName: %v, valid: %v", gang.Name, valid)
+}
+
+func (gang *Gang) setChildScheduleCycle(pod *v1.Pod, childCycle int) {
+	gang.lock.Lock()
+	defer gang.lock.Unlock()
+
+	podId := GetNamespaceSplicingName(pod.Namespace, pod.Name)
+	gang.ChildrenScheduleRoundMap[podId] = childCycle
+}
+
+func (gang *Gang) trySetScheduleCycleValid() {
+	gang.lock.Lock()
+	defer gang.lock.Unlock()
+
+	num := 0
+	for _, childScheduleCycle := range gang.ChildrenScheduleRoundMap {
+		if childScheduleCycle == gang.ScheduleCycle {
+			num++
+		}
+	}
+
+	if num == gang.TotalChildrenNum {
+		gang.ScheduleCycleValid = true
+		gang.ScheduleCycle += 1
+		klog.Infof("trySetScheduleCycleTrue, gangName: %v, ScheduleCycle: %v, ScheduleCycleValid: %v",
+			gang.Name, gang.ScheduleCycle, true)
+	}
+}
+
+func (gang *Gang) addAssumedPod(pod *v1.Pod) {
+	gang.lock.Lock()
+	defer gang.lock.Unlock()
+
+	podId := GetNamespaceSplicingName(pod.Namespace, pod.Name)
+	gang.WaitingForBindChildren[podId] = pod
+	klog.Infof("AddAssumedPod, gangName: %v, podName: %v", gang.Name, podId)
+
+	if len(gang.WaitingForBindChildren)+len(gang.BoundChildren) >= gang.MinRequiredNumber {
+		gang.ResourceSatisfied = true
+		klog.Infof("Gang ResourceSatisfied, gangName: %v", gang.Name)
+	}
+}
+
+func (gang *Gang) getChildrenFromGang() (children []*v1.Pod) {
+	gang.lock.Lock()
+	defer gang.lock.Unlock()
+	children = make([]*v1.Pod, 0)
+	for _, pod := range gang.Children {
+		child := pod
+		children = append(children, child)
+	}
+	return
+}
+
+func (gang *Gang) delAssumedPod(pod *v1.Pod) {
+	gang.lock.Lock()
+	defer gang.lock.Unlock()
+
+	podId := GetNamespaceSplicingName(pod.Namespace, pod.Name)
+	delete(gang.WaitingForBindChildren, podId)
+	klog.Infof("delAssumedPod, gangName: %v, podName: %v", gang.Name, podId)
+
+	if len(gang.WaitingForBindChildren)+len(gang.BoundChildren) < gang.MinRequiredNumber {
+		gang.ResourceSatisfied = false
```

Comment:
once gang satisfied, never go back to false. and I think gang.ResourceSatisfied = true should move to addBoundPod; only bind sucess is can really lead to "ResourceSatisfied"